server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -50,187 +52,267 @@
-            {
-              replicationManager.unregisterReplicantCreation(
-                  tier,
-                  segment.getIdentifier(),
-                  holder.getServer().getHost()
-              );
-            }
-          }
-      );
-
-      ++assignedCount;
-      ++currReplicantsInTier;
-      ++currTotalReplicantsInCluster;
+      final String segmentId = segment.getIdentifier();
+      final String holderHost = holder.getServer().getHost();
+      throttler.registerReplicantCreation(tier, segmentId, holderHost);
+      holder.getPeon().loadSegment(segment, () -> throttler.unregisterReplicantCreation(tier, segmentId, holderHost));
     }
 
-    return assignedCount;
+    return numToAssign;
   }
 
-  private CoordinatorStats drop(
-      final Map<String, Integer> loadStatus,
+  private static void drop(
+      final Object2IntMap<String> targetReplicants,
+      final Object2IntMap<String> currentReplicants,
+      final DruidCoordinatorRuntimeParams params,
       final DataSegment segment,
-      final DruidCoordinatorRuntimeParams params
+      final CoordinatorStats stats
   )
   {
-    CoordinatorStats stats = new CoordinatorStats();
+    final DruidCluster druidCluster = params.getDruidCluster();
 
     // Make sure we have enough loaded replicants in the correct tiers in the cluster before doing anything
-    for (Integer leftToLoad : loadStatus.values()) {
-      if (leftToLoad > 0) {
-        return stats;
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
+      final String tier = entry.getKey();
+      // if there are replicants loading in cluster
+      if (druidCluster.hasTier(tier) && entry.getIntValue() > currentReplicants.getOrDefault(tier, 0)) {
+        return;
       }
     }
 
-    // Find all instances of this segment across tiers
-    Map<String, Integer> replicantsByTier = params.getSegmentReplicantLookup().getClusterTiers(segment.getIdentifier());
-
-    for (Map.Entry<String, Integer> entry : replicantsByTier.entrySet()) {
+    for (final Object2IntMap.Entry<String> entry : currentReplicants.object2IntEntrySet()) {
       final String tier = entry.getKey();
-      int loadedNumReplicantsForTier = entry.getValue();
-      int expectedNumReplicantsForTier = getNumReplicants(tier);
 
-      stats.addToTieredStat(DROPPED_COUNT, tier, 0);
+      final MinMaxPriorityQueue<ServerHolder> holders = druidCluster.getHistoricalsByTier(tier);
 
-      MinMaxPriorityQueue<ServerHolder> serverQueue = params.getDruidCluster().getHistoricalsByTier(tier);
-      if (serverQueue == null) {
-        log.makeAlert("No holders found for tier[%s]", entry.getKey()).emit();
-        continue;
+      final int numDropped;
+      if (holders == null) {
+        log.makeAlert("No holders found for tier[%s]", tier).emit();
+        numDropped = 0;
+      } else {
+        final int numToDrop = entry.getIntValue() - targetReplicants.getOrDefault(tier, 0);
```

Comment:
Suggested to extract `entry.getIntValue()` as "currentReplicants" for readability

Reply:
:+1: